service/r2/routes.go

```diff
@@ -271,3 +275,117 @@ func fetchRollupRuleHistory(s *service, r *http.Request) (data interface{}, err
 	}
 	return models.NewRollupRuleSnapshots(hist), nil
 }
+
+func bulkUpdateRuleSet(s *service, r *http.Request) (data interface{}, err error) {
+	vars := mux.Vars(r)
+	var req bulkRequest
+	if err := parseRequest(&req, r.Body); err != nil {
+		return nil, err
+	}
+	uOpts, err := s.newUpdateOptions(r)
+	if err != nil {
+		return nil, err
+	}
+
+	originalRS, err := s.store.FetchRuleSet(vars[namespaceIDVar])
+	if err != nil {
+		return nil, err
+	}
+	if req.RuleSetVersion != originalRS.Version() {
+		return nil, NewConflictError(
+			fmt.Sprintf("can not update namespace %s, version mismatch", vars[namespaceIDVar]),
+		)
+	}
+
+	newRS, err := applyChangesToRuleSet(
+		req.RuleSetChanges,
+		originalRS.ToMutableRuleSet().Clone(),
+		uOpts,
+		s.updateHelper,
+	)
+	if err != nil {
+		return nil, NewBadInputError(err.Error())
+	}
+
+	rs, err := s.store.UpdateRuleSet(newRS)
+	if err != nil {
+		return nil, err
+	}
+	latest, err := rs.Latest()
+	if err != nil {
+		return nil, err
+	}
+	serializableRuleSet := models.NewRuleSet(latest)
+
+	return serializableRuleSet, nil
+}
+
+func applyChangesToRuleSet(
+	rsc changes.RuleSetChanges,
+	ruleSet rules.MutableRuleSet,
+	uOpts store.UpdateOptions,
+	helper rules.RuleSetUpdateHelper,
+) (rules.MutableRuleSet, error) {
+
+	if len(rsc.MappingRuleChanges) == 0 &&
+		len(rsc.RollupRuleChanges) == 0 {
+		return nil, fmt.Errorf("invalid request: no ruleset changes detected")
+	}
+
+	meta := helper.NewUpdateMetadata(
+		time.Now().Unix(),
+		uOpts.Author(),
+	)
+
+	for _, mrChange := range rsc.MappingRuleChanges {
+		if mrChange.Op == "" {
+			return nil, fmt.Errorf("invalid request: changes must contain an op")
```

Comment:
nit: specify that this is MappingRule in the err string and/or create a new: 
`var errNoOpChage = errors.New(...)`

Also don't need fmt.Errorf

Reply:
Yeah that's a great idea. Initially, there was some formatting in this error messag but it was taken out because there was potential for a panic. Now that there is no formatting the pattern you are suggesting make more sense.